File: hadron-desktop/src-tauri/src/commands/summaries.rs
//! Session Summary Tauri commands for Ask Hadron 2.0

use std::sync::Arc;

use crate::ai_service::{
    build_chat_request_anthropic, build_chat_request_openai, call_provider_chat, ChatMessage,
};
use crate::database::{Database, SessionSummary};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GenerateSummaryRequest {
    pub session_id: String,
    pub provider: String,
    pub model: String,
    pub api_key: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveSummaryRequest {
    pub session_id: String,
    pub summary_markdown: String,
    pub topic: String,
    pub won_version: Option<String>,
    pub customer: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExportSummariesRequest {
    pub date_from: Option<String>,
    pub date_to: Option<String>,
    pub customer: Option<String>,
    pub unexported_only: Option<bool>,
}

#[tauri::command]
pub async fn generate_session_summary(
    db: tauri::State<'_, Arc<Database>>,
    request: GenerateSummaryRequest,
) -> Result<String, String> {
    log::debug!("cmd: generate_session_summary");
    // Load all messages for session
    let messages = db
        .get_chat_messages(&request.session_id)
        .map_err(|e| e.to_string())?;
    if messages.is_empty() {
        return Err("No messages in session".to_string());
    }

    // Build conversation transcript
    let mut transcript = String::new();
    for msg in &messages {
        let role_label = match msg.role.as_str() {
            "user" => "User",
            "assistant" => "Hadron",
            _ => continue,
        };
        transcript.push_str(&format!("**{}:** {}\n\n", role_label, msg.content));
    }

    // Build summarization prompt
    let system_prompt = "You are a technical writer. Summarize the following support conversation into a structured document. Use this exact format:\n\n## Topic\n[One-line description]\n\n## Context\n- WON Version: [if mentioned]\n- Customer: [if mentioned]\n- Related Analyses: [#IDs if any]\n\n## Question\n[The core question or problem]\n\n## Answer\n[Condensed key findings]\n\n## Sources\n[KB docs, analysis IDs, JIRA tickets cited]\n\n## Resolution\n[Action taken or recommended]\n\nBe concise. Only include sections that have content.";

    let chat_messages = vec![ChatMessage {
        role: "user".to_string(),
        content: transcript,
    }];

    let body = match request.provider.as_str() {
        "anthropic" => build_chat_request_anthropic(
            &chat_messages,
            system_prompt,
            &request.model,
            4000,
            false,
        ),
        _ => build_chat_request_openai(
            &chat_messages,
            system_prompt,
            &request.model,
            4000,
            false,
        ),
    };

    let response = call_provider_chat(&request.provider, body, &request.api_key).await?;

    Ok(response.content)
}

#[tauri::command]
pub async fn save_session_summary(
    db: tauri::State<'_, Arc<Database>>,
    request: SaveSummaryRequest,
) -> Result<i64, String> {
    log::debug!("cmd: save_session_summary");
    db.save_session_summary(
        &request.session_id,
        &request.summary_markdown,
        Some(&request.topic),
        request.won_version.as_deref(),
        request.customer.as_deref(),
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_session_summary(
    db: tauri::State<'_, Arc<Database>>,
    session_id: String,
) -> Result<Option<SessionSummary>, String> {
    log::debug!("cmd: get_session_summary");
    db.get_session_summary(&session_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn export_summaries_bundle(
    db: tauri::State<'_, Arc<Database>>,
    request: ExportSummariesRequest,
) -> Result<String, String> {
    log::debug!("cmd: export_summaries_bundle");
    let summaries = db
        .get_summaries_for_export(
            request.date_from.as_deref(),
            request.date_to.as_deref(),
            request.customer.as_deref(),
            request.unexported_only.unwrap_or(false),
        )
        .map_err(|e| e.to_string())?;

    let mut bundle: Vec<serde_json::Value> = Vec::new();
    for summary in &summaries {
        // Build a sanitized filename from date + topic
        let date_part = summary
            .created_at
            .split(' ')
            .next()
            .unwrap_or("unknown-date");
        let topic_part = summary
            .topic
            .as_deref()
            .unwrap_or("untitled")
            .to_lowercase()
            .replace(' ', "-")
            .chars()
            .filter(|c| c.is_alphanumeric() || *c == '-')
            .take(50)
            .collect::<String>();
        let filename = format!("{}-{}.md", date_part, topic_part);

        bundle.push(serde_json::json!({
            "filename": filename,
            "content": summary.summary_markdown,
        }));
    }

    serde_json::to_string(&bundle).map_err(|e| e.to_string())
}

File: hadron-electron/electron/ipc/summaries.ts
import { IpcMain } from 'electron'
import { getDb } from '../database'
import { callAi } from '../services/ai-service'
import { getSecret } from '../services/safe-storage'
import { SERVICE_NAME } from '../services/jira-client'
import { wrapField } from '../services/prompt-helpers'

const SUMMARY_SYSTEM_PROMPT = `You are a technical writer. Summarize the following support conversation into a structured document. Use this exact format:

## Topic
[One-line description]

## Context
- WON Version: [if mentioned]
- Customer: [if mentioned]

## Question
[The core question or problem]

## Answer
[Condensed key findings]

## Resolution
[Action taken or recommended]

Be concise. Only include sections that have content.`

export function registerSummaryHandlers(ipcMain: IpcMain): void {
  // Frontend wraps all write calls in { request: params } — unwrap before accessing fields.
  ipcMain.handle('generate_session_summary', async (_e, args: {
    request?: { sessionId: string; provider: string; model: string; apiKey?: string }
    sessionId?: string; provider?: string; model?: string; apiKey?: string
  }) => {
    const p = args.request ?? (args as { sessionId: string; provider: string; model: string; apiKey?: string })
    const db = getDb()
    const messages = db.prepare(
      'SELECT role, content FROM chat_messages WHERE session_id = ? ORDER BY timestamp ASC'
    ).all(p.sessionId) as Array<{ role: string; content: string }>

    if (messages.length === 0) throw new Error('No messages in session')

    const transcript = messages
      .filter(m => m.role === 'user' || m.role === 'assistant')
      .map(m => `**${m.role === 'user' ? 'User' : 'Hadron'}:** ${wrapField('MESSAGE', m.content)}`)
      .join('\n\n')

    let apiKey = p.apiKey ?? ''
    if (!apiKey) apiKey = getSecret(SERVICE_NAME, p.provider ?? '') ?? ''
    if (!apiKey) throw new Error(`No API key for provider: ${p.provider}`)

    const result = await callAi({
      provider: p.provider ?? '',
      model: p.model ?? '',
      apiKey,
      systemPrompt: SUMMARY_SYSTEM_PROMPT,
      userPrompt: transcript,
      maxTokens: 4000,
    })
    return result.content
  })

  ipcMain.handle('save_session_summary', (_e, args: {
    request?: { sessionId: string; summaryMarkdown: string; topic: string; wonVersion?: string; customer?: string }
    sessionId?: string; summaryMarkdown?: string; topic?: string; wonVersion?: string; customer?: string
  }) => {
    const p = args.request ?? (args as { sessionId: string; summaryMarkdown: string; topic: string; wonVersion?: string; customer?: string })
    const db = getDb()
    const now = new Date().toISOString()
    const existing = db.prepare(
      'SELECT id FROM session_summaries WHERE session_id = ?'
    ).get(p.sessionId) as { id: number } | undefined

    if (existing) {
      db.prepare(`UPDATE session_summaries
        SET summary_markdown = ?, topic = ?, won_version = ?, customer = ?, updated_at = ?
        WHERE id = ?`)
        .run(p.summaryMarkdown, p.topic, p.wonVersion ?? null, p.customer ?? null, now, existing.id)
      return existing.id
    }
    const row = db.prepare(`INSERT INTO session_summaries
      (session_id, summary_markdown, topic, won_version, customer, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `).run(p.sessionId, p.summaryMarkdown, p.topic,
           p.wonVersion ?? null, p.customer ?? null, now, now)
    return row.lastInsertRowid
  })

  ipcMain.handle('get_session_summary', (_e, args: { sessionId: string }) => {
    const db = getDb()
    return db.prepare('SELECT * FROM session_summaries WHERE session_id = ?').get(args.sessionId) ?? null
  })

  ipcMain.handle('export_summaries_bundle', (_e, args: {
    request?: { dateFrom?: string; dateTo?: string; customer?: string; unexportedOnly?: boolean }
    dateFrom?: string; dateTo?: string; customer?: string; unexportedOnly?: boolean
  }) => {
    const p = args.request ?? args
    const db = getDb()
    let sql = 'SELECT * FROM session_summaries WHERE 1=1'
    const params: unknown[] = []
    if (p.dateFrom) { sql += ' AND created_at >= ?'; params.push(p.dateFrom) }
    if (p.dateTo)   { sql += ' AND created_at <= ?'; params.push(p.dateTo) }
    if (p.customer) { sql += ' AND customer = ?'; params.push(p.customer) }
    if (p.unexportedOnly) { sql += ' AND is_exported = 0' }
    const summaries = db.prepare(sql).all(...params) as Array<Record<string, unknown>>

    const bundle = summaries.map(s => {
      const datePart = (s.created_at as string).split(' ')[0] ?? 'unknown-date'
      const topicPart = (s.topic as string ?? 'untitled').toLowerCase().replace(/ /g, '-')
        .replace(/[^a-z0-9-]/g, '').substring(0, 50)
      return { filename: `${datePart}-${topicPart}.md`, content: s.summary_markdown }
    })
    return JSON.stringify(bundle)
  })
}
